File: Kalorie/Kalorie/Features/Dashboard/DashboardConfigurator.swift
//
//  DashboardConfigurator.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.06.2026.
//

import Foundation

struct DashboardConfigurator {

    // MARK: - Functions

    func createView(mergeStatusReporting: any MergeStatusReporting) -> DashboardView {
        let dataProvider = FirestoreDataProvider()
        let authProvider = AuthProvider()
        return DashboardView(
            viewModel: DashboardViewModel(
                fetchMealTypes: FetchMealTypesUseCase(dataProvider: dataProvider, authProvider: authProvider),
                fetchFoodsConsumedForMonth: FetchFoodsConsumedForMonthUseCase(dataProvider: dataProvider, authProvider: authProvider),
                setupDefaultMeals: SetupDefaultMealsUseCase(dataProvider: dataProvider, authProvider: authProvider),
                confirmMealTypesEmpty: ConfirmMealTypesEmptyUseCase(dataProvider: dataProvider, authProvider: authProvider),
                deleteFoodConsumed: DeleteFoodConsumedUseCase(dataProvider: dataProvider, authProvider: authProvider)
            ),
            router: DashboardRouter(
                mealTypeSheetConfigurator: MealTypeSheetConfigurator(),
                addFoodSheetConfigurator: AddFoodSheetConfigurator(dataProvider: dataProvider, authProvider: authProvider),
                foodConsumedDetailConfigurator: FoodConsumedDetailConfigurator(dataProvider: dataProvider, authProvider: authProvider),
                accountConfigurator: AccountConfigurator(dataProvider: dataProvider, authProvider: authProvider, mergeStatusReporting: mergeStatusReporting),
                myCreatedMealEditorConfigurator: MyCreatedMealEditorConfigurator(dataProvider: dataProvider, authProvider: authProvider)
            )
        )
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/dashboard/DashboardConfigurator.kt
package antoni.kalorie.features.dashboard

import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.viewmodel.compose.viewModel
import antoni.kalorie.R
import antoni.kalorie.core.auth.AuthProvider
import antoni.kalorie.core.networking.FirestoreDataProvider
import antoni.kalorie.core.usecases.ConfirmMealTypesEmptyUseCase
import antoni.kalorie.core.usecases.DeleteFoodConsumedUseCase
import antoni.kalorie.core.usecases.FetchFoodsConsumedForMonthUseCase
import antoni.kalorie.core.usecases.FetchMealTypesUseCase
import antoni.kalorie.core.usecases.SetupDefaultMealsUseCase

class DashboardConfigurator {

    // MARK: - Functions

    @Composable
    fun createView(userId: String?) {
        val mealNames = listOf(
            stringResource(R.string.defaultMeals_breakfast),
            stringResource(R.string.defaultMeals_secondBreakfast),
            stringResource(R.string.defaultMeals_lunch),
            stringResource(R.string.defaultMeals_snack),
            stringResource(R.string.defaultMeals_dinner),
        )
        val viewModel = viewModel(key = userId) {
            val dataProvider = FirestoreDataProvider()
            val authProvider = AuthProvider()
            DashboardViewModel(
                fetchMealTypes = FetchMealTypesUseCase(dataProvider, authProvider),
                fetchFoodsConsumedForMonth = FetchFoodsConsumedForMonthUseCase(dataProvider, authProvider),
                setupDefaultMeals = SetupDefaultMealsUseCase(dataProvider, authProvider, mealNames),
                confirmMealTypesEmpty = ConfirmMealTypesEmptyUseCase(dataProvider, authProvider),
                deleteFoodConsumed = DeleteFoodConsumedUseCase(dataProvider, authProvider),
            )
        }
        DashboardView(viewModel = viewModel)
    }
}
